File: backend/go/internal/services/bridge_service.go
package services

import (
	"context"
	"database/sql"
	"fmt"
	"time"

	"github.com/google/uuid"
	"github.com/tigerwallet/backend/internal/models"
)

type BridgeService struct {
	db *sql.DB
}

func NewBridgeService(db *sql.DB) *BridgeService {
	return &BridgeService{db: db}
}

// Supported chains
var SupportedChains = map[string]struct{}{
	"ETHEREUM":    {},
	"BSC":         {},
	"POLYGON":     {},
	"AVALANCHE":   {},
	"ARBITRUM":    {},
	"OPTIMISM":    {},
	"SOLANA":      {},
	"NEAR":        {},
	"APTOS":       {},
	"SUI":         {},
}

// Initiate bridge transaction
func (s *BridgeService) InitiateBridge(ctx context.Context, userID uuid.UUID, fromChain, toChain string, token string, amount float64) (*models.BridgeTransaction, error) {
	// Validate chains
	if _, ok := SupportedChains[fromChain]; !ok {
		return nil, fmt.Errorf("unsupported source chain: %s", fromChain)
	}
	if _, ok := SupportedChains[toChain]; !ok {
		return nil, fmt.Errorf("unsupported destination chain: %s", toChain)
	}

	if fromChain == toChain {
		return nil, fmt.Errorf("source and destination chains must be different")
	}

	// Get bridge fee
	fee, err := s.getBridgeFee(fromChain, toChain, token)
	if err != nil {
		return nil, err
	}

	// Create bridge transaction
	tx := &models.BridgeTransaction{
		ID:              uuid.New(),
		UserID:          userID,
		FromChain:       fromChain,
		ToChain:         toChain,
		Token:           token,
		Amount:          amount,
		Fee:             fee,
		ReceivedAmount:  amount - fee,
		Status:          "PENDING",
		SourceTxHash:    "",
		DestinationTxHash: "",
		CreatedAt:       time.Now(),
		UpdatedAt:       time.Now(),
	}

	_, err = s.db.ExecContext(ctx, `
		INSERT INTO bridge_transactions 
		(id, user_id, from_chain, to_chain, token, amount, fee, received_amount, status, created_at, updated_at)
		VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11)
	`, tx.ID, tx.UserID, tx.FromChain, tx.ToChain, tx.Token, tx.Amount, tx.Fee, tx.ReceivedAmount, tx.Status, tx.CreatedAt, tx.UpdatedAt)

	if err != nil {
		return nil, fmt.Errorf("failed to create bridge transaction: %w", err)
	}

	return tx, nil
}

// Confirm source chain deposit
func (s *BridgeService) confirmSourceDeposit(ctx context.Context, txID uuid.UUID, sourceTxHash string) error {
	result, err := s.db.ExecContext(ctx, `
		UPDATE bridge_transactions 
		SET status = 'CONFIRMED', source_tx_hash = $2, updated_at = NOW()
		WHERE id = $1 AND status = 'PENDING'
	`, txID, sourceTxHash)

	if err != nil {
		return fmt.Errorf("failed to confirm deposit: %w", err)
	}

	rows, _ := result.RowsAffected()
	if rows == 0 {
		return fmt.Errorf("transaction not found or already processed")
	}

	return nil
}

// Complete bridge on destination chain
func (s *BridgeService) CompleteBridge(ctx context.Context, txID uuid.UUID, destTxHash string) error {
	result, err := s.db.ExecContext(ctx, `
		UPDATE bridge_transactions 
		SET status = 'COMPLETED', destination_tx_hash = $2, updated_at = NOW()
		WHERE id = $1 AND status = 'CONFIRMED'
	`, txID, destTxHash)

	if err != nil {
		return fmt.Errorf("failed to complete bridge: %w", err)
	}

	rows, _ := result.RowsAffected()
	if rows == 0 {
		return fmt.Errorf("transaction not found or not confirmed")
	}

	return nil
}

// Cancel bridge
func (s *BridgeService) CancelBridge(ctx context.Context, txID uuid.UUID) error {
	result, err := s.db.ExecContext(ctx, `
		UPDATE bridge_transactions 
		SET status = 'CANCELLED', updated_at = NOW()
		WHERE id = $1 AND status = 'PENDING'
	`, txID)

	if err != nil {
		return fmt.Errorf("failed to cancel bridge: %w", err)
	}

	rows, _ := result.RowsAffected()
	if rows == 0 {
		return fmt.Errorf("transaction not found or already processed")
	}

	return nil
}

// Get user bridge transactions
func (s *BridgeService) GetUserTransactions(ctx context.Context, userID uuid.UUID) ([]models.BridgeTransaction, error) {
	rows, err := s.db.QueryContext(ctx, `
		SELECT id, user_id, from_chain, to_chain, token, amount, fee, received_amount, 
			   status, source_tx_hash, destination_tx_hash, created_at, updated_at
		FROM bridge_transactions 
		WHERE user_id = $1
		ORDER BY created_at DESC
	`, userID)

	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var txs []models.BridgeTransaction
	for rows.Next() {
		var tx models.BridgeTransaction
		err := rows.Scan(&tx.ID, &tx.UserID, &tx.FromChain, &tx.ToChain, &tx.Token, 
			&tx.Amount, &tx.Fee, &tx.ReceivedAmount, &tx.Status, 
			&tx.SourceTxHash, &tx.DestinationTxHash, &tx.CreatedAt, &tx.UpdatedAt)
		if err != nil {
			continue
		}
		txs = append(txs, tx)
	}

	return txs, nil
}

// Get bridge status
func (s *BridgeService) GetTransactionStatus(ctx context.Context, txID uuid.UUID) (*models.BridgeTransaction, error) {
	var tx models.BridgeTransaction
	err := s.db.QueryRowContext(ctx, `
		SELECT id, user_id, from_chain, to_chain, token, amount, fee, received_amount, 
			   status, source_tx_hash, destination_tx_hash, created_at, updated_at
		FROM bridge_transactions 
		WHERE id = $1
	`, txID).Scan(&tx.ID, &tx.UserID, &tx.FromChain, &tx.ToChain, &tx.Token, 
		&tx.Amount, &tx.Fee, &tx.ReceivedAmount, &tx.Status, 
		&tx.SourceTxHash, &tx.DestinationTxHash, &tx.CreatedAt, &tx.UpdatedAt)

	if err == sql.ErrNoRows {
		return nil, fmt.Errorf("transaction not found")
	}

	return &tx, err
}

// Get supported tokens for bridge
func (s *BridgeService) GetSupportedTokens(ctx context.Context, chain string) ([]models.BridgeToken, error) {
	rows, err := s.db.QueryContext(ctx, `
		SELECT token, chain, min_amount, max_amount, is_active
		FROM bridge_tokens
		WHERE chain = $1 AND is_active = true
	`, chain)

	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var tokens []models.BridgeToken
	for rows.Next() {
		var t models.BridgeToken
		err := rows.Scan(&t.Token, &t.Chain, &t.MinAmount, &t.MaxAmount, &t.IsActive)
		if err != nil {
			continue
		}
		tokens = append(tokens, t)
	}

	return tokens, nil
}

// Get bridge fee
func (s *BridgeService) getBridgeFee(fromChain, toChain, token string) (float64, error) {
	var fee float64
	err := s.db.QueryRow(`
		SELECT fee_percentage FROM bridge_fees 
		WHERE from_chain = $1 AND to_chain = $2 AND token = $3
	`, fromChain, toChain, token).Scan(&fee)

	if err == sql.ErrNoRows {
		// Default 0.3%
		return 0.003, nil
	}

	return fee, err
}

// Get estimated receive amount
func (s *BridgeService) EstimateReceive(ctx context.Context, fromChain, toChain string, token string, amount float64) (float64, float64, error) {
	fee, err := s.getBridgeFee(fromChain, toChain, token)
	if err != nil {
		return 0, 0, err
	}

	received := amount * (1 - fee)
	return received, fee, nil
}

// Get bridge statistics
func (s *BridgeService) GetBridgeStats(ctx context.Context) (map[string]interface{}, error) {
	var totalVolume, totalTransactions float64

	s.db.QueryRow(`
		SELECT COALESCE(SUM(amount), 0), COALESCE(COUNT(*), 0)
		FROM bridge_transactions
		WHERE status = 'COMPLETED'
	`).Scan(&totalVolume, &totalTransactions)

	stats := map[string]interface{}{
		"totalVolume":       totalVolume,
		"totalTransactions": totalTransactions,
		"supportedChains":   len(SupportedChains),
	}

	return stats, nil
}

File: mobile/flutter/lib/features/bridge/bridge_service.dart
// Bridge Service - Flutter Mobile
// Cross-chain bridging with real backend

import 'dart:convert';
import 'package:http/http.dart' as http;

class BridgeService {
  static const String API_BASE = 'https://api.tigerwallet.com/api/v1';
  String? _token;
  
  BridgeService({String? token}) : _token = token;
  
  Map<String, String> get _headers => {
    'Content-Type': 'application/json',
    if (_token != null) 'Authorization': 'Bearer $_token',
  };
  
  // Get supported chains
  Future<List<Chain>> getSupportedChains() async {
    final response = await http.get(
      Uri.parse('$API_BASE/bridge/chains'),
      headers: _headers,
    );
    
    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      return (data['data'] as List).map((c) => Chain.fromJson(c)).toList();
    }
    throw Exception('Failed to load chains');
  }
  
  // Get supported tokens for a chain
  Future<List<BridgeToken>> getTokens(String chain) async {
    final response = await http.get(
      Uri.parse('$API_BASE/bridge/tokens?chain=$chain'),
      headers: _headers,
    );
    
    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      return (data['data'] as List).map((t) => BridgeToken.fromJson(t)).toList();
    }
    throw Exception('Failed to load tokens');
  }
  
  // Get estimated receive amount
  Future<BridgeEstimate> getEstimate({
    required String fromChain,
    required String toChain,
    required String token,
    required double amount,
  }) async {
    final response = await http.get(
      Uri.parse(
        '$API_BASE/bridge/estimate?fromChain=$fromChain&toChain=$toChain&token=$token&amount=$amount',
      ),
      headers: _headers,
    );
    
    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      return BridgeEstimate.fromJson(data['data']);
    }
    throw Exception('Failed to get estimate');
  }
  
  // Initiate bridge transaction
  Future<BridgeTransaction> initiateBridge({
    required String fromChain,
    required String toChain,
    required String token,
    required double amount,
  }) async {
    final response = await http.post(
      Uri.parse('$API_BASE/bridge/transactions'),
      headers: _headers,
      body: json.encode({
        'fromChain': fromChain,
        'toChain': toChain,
        'token': token,
        'amount': amount,
      }),
    );
    
    if (response.statusCode == 201) {
      final data = json.decode(response.body);
      return BridgeTransaction.fromJson(data['data']);
    }
    throw Exception('Failed to initiate bridge');
  }
  
  // Confirm source deposit (after user sends tokens)
  Future<bool> confirmDeposit(String txId, String sourceTxHash) async {
    final response = await http.post(
      Uri.parse('$API_BASE/bridge/transactions/$txId/confirm'),
      headers: _headers,
      body: json.encode({'sourceTxHash': sourceTxHash}),
    );
    
    return response.statusCode == 200;
  }
  
  // Complete bridge (after bridge completes on dest chain)
  Future<bool> completeBridge(String txId, String destTxHash) async {
    final response = await http.post(
      Uri.parse('$API_BASE/bridge/transactions/$txId/complete'),
      headers: _headers,
      body: json.encode({'destTxHash': destTxHash}),
    );
    
    return response.statusCode == 200;
  }
  
  // Cancel bridge
  Future<bool> cancelBridge(String txId) async {
    final response = await http.post(
      Uri.parse('$API_BASE/bridge/transactions/$txId/cancel'),
      headers: _headers,
    );
    
    return response.statusCode == 200;
  }
  
  // Get transaction status
  Future<BridgeTransaction> getTransactionStatus(String txId) async {
    final response = await http.get(
      Uri.parse('$API_BASE/bridge/transactions/$txId'),
      headers: _headers,
    );
    
    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      return BridgeTransaction.fromJson(data['data']);
    }
    throw Exception('Failed to get transaction');
  }
  
  // Get user transactions
  Future<List<BridgeTransaction>> getUserTransactions() async {
    final response = await http.get(
      Uri.parse('$API_BASE/bridge/transactions'),
      headers: _headers,
    );
    
    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      return (data['data'] as List).map((t) => BridgeTransaction.fromJson(t)).toList();
    }
    return [];
  }
  
  // Get bridge statistics
  Future<BridgeStats> getStats() async {
    final response = await http.get(
      Uri.parse('$API_BASE/bridge/stats'),
      headers: _headers,
    );
    
    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      return BridgeStats.fromJson(data['data']);
    }
    throw Exception('Failed to load stats');
  }
}

class Chain {
  final String id;
  final String name;
  final String symbol;
  final String icon;
  final bool isActive;
  final List<String> features;
  
  Chain({
    required this.id,
    required this.name,
    required this.symbol,
    required this.icon,
    required this.isActive,
    required this.features,
  });
  
  factory Chain.fromJson(Map<String, dynamic> json) {
    return Chain(
      id: json['id'] ?? '',
      name: json['name'] ?? '',
      symbol: json['symbol'] ?? '',
      icon: json['icon'] ?? '',
      isActive: json['isActive'] ?? false,
      features: List<String>.from(json['features'] ?? []),
    );
  }
}

class BridgeToken {
  final String token;
  final String name;
  final String symbol;
  final String icon;
  final double minAmount;
  final double maxAmount;
  final bool isActive;
  
  BridgeToken({
    required this.token,
    required this.name,
    required this.symbol,
    required this.icon,
    required this.minAmount,
    required this.maxAmount,
    required this.isActive,
  });
  
  factory BridgeToken.fromJson(Map<String, dynamic> json) {
    return BridgeToken(
      token: json['token'] ?? '',
      name: json['name'] ?? '',
      symbol: json['symbol'] ?? '',
      icon: json['icon'] ?? '',
      minAmount: (json['minAmount'] ?? 0).toDouble(),
      maxAmount: (json['maxAmount'] ?? 0).toDouble(),
      isActive: json['isActive'] ?? false,
    );
  }
}

class BridgeEstimate {
  final double receivedAmount;
  final double fee;
  final double feePercentage;
  final String estimatedTime;
  
  BridgeEstimate({
    required this.receivedAmount,
    required this.fee,
    required this.feePercentage,
    required this.estimatedTime,
  });
  
  factory BridgeEstimate.fromJson(Map<String, dynamic> json) {
    return BridgeEstimate(
      receivedAmount: (json['receivedAmount'] ?? 0).toDouble(),
      fee: (json['fee'] ?? 0).toDouble(),
      feePercentage: (json['feePercentage'] ?? 0).toDouble(),
      estimatedTime: json['estimatedTime'] ?? '',
    );
  }
}

class BridgeTransaction {
  final String id;
  final String fromChain;
  final String toChain;
  final String token;
  final double amount;
  final double fee;
  final double receivedAmount;
  final String status;
  final String? sourceTxHash;
  final String? destTxHash;
  final DateTime createdAt;
  final DateTime? updatedAt;
  
  BridgeTransaction({
    required this.id,
    required this.fromChain,
    required this.toChain,
    required this.token,
    required this.amount,
    required this.fee,
    required this.receivedAmount,
    required this.status,
    this.sourceTxHash,
    this.destTxHash,
    required this.createdAt,
    this.updatedAt,
  });
  
  factory BridgeTransaction.fromJson(Map<String, dynamic> json) {
    return BridgeTransaction(
      id: json['id'] ?? '',
      fromChain: json['fromChain'] ?? '',
      toChain: json['toChain'] ?? '',
      token: json['token'] ?? '',
      amount: (json['amount'] ?? 0).toDouble(),
      fee: (json['fee'] ?? 0).toDouble(),
      receivedAmount: (json['receivedAmount'] ?? 0).toDouble(),
      status: json['status'] ?? 'PENDING',
      sourceTxHash: json['sourceTxHash'],
      destTxHash: json['destTxHash'],
      createdAt: DateTime.parse(json['createdAt']),
      updatedAt: json['updatedAt'] != null ? DateTime.parse(json['updatedAt']) : null,
    );
  }
}

class BridgeStats {
  final double totalVolume;
  final int totalTransactions;
  final int supportedChains;
  final int supportedTokens;
  
  BridgeStats({
    required this.totalVolume,
    required this.totalTransactions,
    required this.supportedChains,
    required this.supportedTokens,
  });
  
  factory BridgeStats.fromJson(Map<String, dynamic> json) {
    return BridgeStats(
      totalVolume: (json['totalVolume'] ?? 0).toDouble(),
      totalTransactions: json['totalTransactions'] ?? 0,
      supportedChains: json['supportedChains'] ?? 0,
      supportedTokens: json['supportedTokens'] ?? 0,
    );
  }
}
